nat-lab/bin/netcat.py

```diff
@@ -0,0 +1,209 @@
+            self.sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+
+        port = self.args.port
+        bind_addr = "::" if self.ipv6 else "0.0.0.0"
+        self.sock.bind((bind_addr, port))
+
+        if self.udp:
+            self._vprint(f"Bound on {bind_addr} {port}")
+        else:
+            self.sock.listen(1)
+            self._vprint(f"Listening on {bind_addr} {port}")
+            conn, addr = self.sock.accept()
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.sock = conn
+
+    def _udptest(self):
+        """Test UDP connection"""
+        # Try sending data to the socket
+        for _ in range(UDP_SCAN_COUNT):
+            self.sock.send(b"X")
+            time.sleep(1)
+
+    def _register_socket(self):
+        """Register the socket to the selector for read events"""
+        self.selector.register(self.sock, selectors.EVENT_READ, self._read_from_socket)
+
+    def _register_stdin(self):
+        """Register stdin to the selector for read events, unless disabled"""
+        if not self.args.d:
+            self.selector.register(
+                sys.stdin, selectors.EVENT_READ, self._read_from_stdin
+            )
+
+    def _read_from_socket(self):
+        """Handle incoming data from the socket"""
+        data, addr = self.sock.recvfrom(RECV_SIZE)
+        if self.udp and self.listen and not self.client_addr:
+            self._vprint(f"Connection received on {addr[0]} {addr[1]}")
+            self.client_addr = addr
+            self.sock.connect(addr)
+        if data:
+            sys.stdout.buffer.write(data)
+            sys.stdout.flush()
+        else:
+            # Connection closed
+            self.selector.unregister(self.sock)
+            self.should_close = True
+
+    def _read_from_stdin(self):
+        """Handle input from stdin and send it over the socket"""
+        data = sys.stdin.buffer.readline()
+        if data:
+            if self.listen and self.client_addr:
+                self.sock.sendto(data, self.client_addr)
+            else:
+                self.sock.send(data)
+        else:
+            # EOF on stdin
+            self.selector.unregister(sys.stdin)
+            self.should_close = True
+
+    def run(self):
+        try:
+            if self.args.l:
+                self._listen()
+            else:
+                self._connect()
+            if not self.args.z:
+                self._register_socket()
+                self._register_stdin()
+                while not self.should_close:
+                    events = self.selector.select(timeout=None)
+                    for key, _ in events:
+                        callback = key.data
+                        callback()
+        except KeyboardInterrupt:
+            print("")
+            sys.exit(2)
+        except OSError as e:
+            if e.errno in (errno.EBADF, errno.ECONNREFUSED):
+                # connection closed
+                sys.exit(0)
```

Comment:
This will introduce silent failures, where we expect NC to work, but for some reason it does not. 

Is there any reason why this condition should be silenced and then process have to exit with success?

Reply:
I'll double check this, I might have over compensated for the other EBADF in connect 😰

Reply:
Right, this actually covered for the case when UDP listener was gone

Reply:
Moved the error log above, so it always logs the OSError